testing/endtoend/evaluators/beaconapi_evaluators/beacon_api.go

```diff
@@ -0,0 +1,433 @@
+				blockP,
+				blockL)
+		}
+	}
+	blockheaderData := beaconPathsAndObjects["/beacon/headers/{param1}"]
+	prysmHeader, ok := blockheaderData.prysmResps["json"].(*apimiddleware.BlockHeaderResponseJson)
+	if !ok {
+		return errors.New("failed to cast type")
+	}
+	proposerdutiesData := beaconPathsAndObjects["/validator/duties/proposer/{param1}"]
+	prysmDuties, ok := proposerdutiesData.prysmResps["json"].(*apimiddleware.ProposerDutiesResponseJson)
+	if !ok {
+		return errors.New("failed to cast type")
+	}
+	if prysmHeader.Data.Root != prysmDuties.DependentRoot {
+		fmt.Printf("current slot: %v\n", slots.CurrentSlot(uint64(genesisData.Data.GenesisTime.AsTime().Unix())))
+		return fmt.Errorf("header root %s does not match duties root %s ", prysmHeader.Data.Root, prysmDuties.DependentRoot)
+	}
+
+	return nil
+}
+
+func compareJSONMulticlient(beaconNodeIdx int, base string, path string, respJSONPrysm interface{}, respJSONLighthouse interface{}, customEvaluator func(interface{}, interface{}) error) error {
+	if err := doMiddlewareJSONGetRequest(
+		base,
+		path,
+		beaconNodeIdx,
+		respJSONPrysm,
+	); err != nil {
+		return errors.Wrap(err, "could not perform GET request for Prysm JSON")
+	}
+
+	if err := doMiddlewareJSONGetRequest(
+		base,
+		path,
+		beaconNodeIdx,
+		respJSONLighthouse,
+		"lighthouse",
+	); err != nil {
+		return errors.Wrap(err, "lighthouse json error")
+	}
+	if customEvaluator != nil {
+		return customEvaluator(respJSONPrysm, respJSONLighthouse)
+	} else {
+		return compareJSONResponseObjects(respJSONPrysm, respJSONLighthouse)
+	}
+}
+
+func compareSSZMulticlient(beaconNodeIdx int, base string, path string) ([]byte, []byte, error) {
+	sszrspL, err := doMiddlewareSSZGetRequest(
+		base,
+		path,
+		beaconNodeIdx,
+		"lighthouse",
+	)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "lighthouse json error")
+	}
+
+	sszrspP, err := doMiddlewareSSZGetRequest(
+		base,
+		path,
+		beaconNodeIdx,
+	)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "prysm json error")
+	}
+	if !bytes.Equal(sszrspL, sszrspP) {
+		return nil, nil, errors.New("prysm ssz response does not match lighthouse ssz response")
+	}
+	return sszrspP, sszrspL, nil
+}
+
+func compareJSONResponseObjects(prysmResp interface{}, lighthouseResp interface{}) error {
+	if !reflect.DeepEqual(prysmResp, lighthouseResp) {
+		p, err := json.Marshal(prysmResp)
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal Prysm response to JSON")
+		}
+		l, err := json.Marshal(lighthouseResp)
+		if err != nil {
+			return errors.Wrap(err, "failed to marshal Lighthouse response to JSON")
+		}
+		return fmt.Errorf("prysm response %s does not match lighthouse response %s",
+			string(p),
+			string(l))
+	}
+	return nil
+}
+
+func pathFromParams(path string, params []string) string {
+	var apiPath string
+	for index := range params {
+		apiPath = strings.Replace(path, fmt.Sprintf("{param%d}", index+1), params[index], 1)
+	}
+	if apiPath == "" {
+		return path
+	}
```

Comment:
Shouldn't this be an error condition? It seems to me that someone passed in a malformed path.

Reply:
It should return the input if there are no parameters to replace. previously it just returned the api Path. the call will fail but that shouldn't mean this function should error.

Reply:
Ah, right. In that case I think you can make this code a little simpler by declaring `apiPath := path` first, and then simply returning `apiPath` after the loop.